File: Torch/ValidConvolution.lua
-- Copyright 2016-present, Facebook, Inc.
-- All rights reserved.
--
-- This source code is licensed under the license found in the
-- LICENSE file in the root directory of this source tree.

return function(sparseconvnet)
  local C = sparseconvnet.C

  local Convolution, parent = torch.class(
    'sparseconvnet.ValidConvolution', 'nn.Module', sparseconvnet)

  function Convolution:__init(dimension, nInputPlanes, nOutputPlanes,
      filterSize, bias)
    parent.__init(self)
    self.dimension = dimension
    self.nInputPlanes = nInputPlanes
    self.nOutputPlanes = nOutputPlanes
    self.filterSize = sparseconvnet.toLongTensor(filterSize,dimension)
    self.filterStride = sparseconvnet.toLongTensor(1,dimension)
    self.filterVolume = self.filterSize:prod()
    for i = 1, dimension do
      assert(self.filterSize[i]%2==1)
    end
    self.weight = torch.Tensor(nInputPlanes*self.filterVolume,nOutputPlanes)
    self.gradWeight = torch.Tensor(nInputPlanes*self.filterVolume,nOutputPlanes)
    if (type(bias) ~= 'boolean') or bias then
      self.bias = torch.Tensor(nOutputPlanes)
      self.gradBias = torch.Tensor(nOutputPlanes)
    end
    self.output = {
      features = torch.Tensor(),
    }
    self.gradInput = {
      features = torch.Tensor()
    }
    self:reset()
  end

  function Convolution:reset()
    local stdv = math.sqrt(2/self.nInputPlanes/self.filterVolume)
    self.weight:normal(0, stdv)
    if self.bias then
      self.bias:zero()
    end
    return self
  end

  function Convolution:updateOutput(input)
    assert(input.features:size(2)==self.nInputPlanes)
    self.output.metadata = input.metadata
    self.output.spatialSize = input.spatialSize
    self.shared.forwardPassMultiplyAddCount=
    self.shared.forwardPassMultiplyAddCount+
    C.dimTypedFn(self.dimension, self._type, 'ValidConvolution_updateOutput')(
      input.spatialSize:cdata(),
      self.filterSize:cdata(),
      input.metadata.ffi,
      input.features:cdata(),
      self.output.features:cdata(),
      self.weight:cdata(),
      self.bias and self.bias:cdata(),
      self.filterVolume,
      self.shared.rulesBuffer and self.shared.rulesBuffer:cdata())
    self.shared.forwardPassHiddenStates=
    self.shared.forwardPassHiddenStates + self.output.features:nElement()
    return self.output
  end

  function Convolution:backward(input, gradOutput)
    C.dimTypedFn(self.dimension, self._type, 'ValidConvolution_backward')(
      input.spatialSize:cdata(),
      self.filterSize:cdata(),
      input.metadata.ffi,
      input.features:cdata(),
      self.gradInput.features:cdata(),
      gradOutput.features:cdata(),
      self.weight:cdata(),
      self.gradWeight:cdata(),
      self.gradBias and self.gradBias:cdata() ,
      self.filterVolume,
      self.shared.rulesBuffer and self.shared.rulesBuffer:cdata())
    return self.gradInput
  end

  function Convolution:type(type,tensorCache)
    if type==nil then
      return self._type
    else
      self._type=type
      self.weight = self.weight:type(type)
      self.gradWeight =self.gradWeight:type(type)
      if self.bias then
        self.bias = self.bias:type(type)
        self.gradBias =self.gradBias:type(type)
      end
      self.output.features=self.output.features:type(type)
      self.gradInput.features=self.gradInput.features:type(type)
    end
  end

  function Convolution:__tostring()
    local s = 'ValidConvolution ' .. self.nInputPlanes .. '->' .. self.nOutputPlanes..' C'
    if self.filterSize:max()==self.filterSize:min() and
    self.filterStride:max()==self.filterStride:min() then
      s=s..self.filterSize[1] ..(self.filterStride[1]==1 and
        '' or '/'..self.filterStride[1])
    else
      s=s..'('..self.filterSize[1]
      for i=2,self.dimension do
        s=s..','..self.filterSize[i]
      end
      s=s..')/('..self.filterStride[1]
      for i=2,self.dimension do
        s=s..','..self.filterStride[i]
      end
      s=s..')'
    end
    return s
  end

  function Convolution:clearState()
    self.output={features=self.output.features:set()}
    self.gradInput={features=self.gradInput.features:set()}
    self.rules=nil
  end

  function Convolution:suggestInputSize(nOut)
    return nOut
  end
end

File: PyTorch/sparseconvnet/legacy/validConvolution.py
# Copyright 2016-present, Facebook, Inc.
# All rights reserved.
#
# This source code is licensed under the license found in the
# LICENSE file in the root directory of this source tree.

import torch
import sparseconvnet as s
from . import SparseModule
from ..utils import toLongTensor, dim_typed_fn, optionalTensor, nullptr
from .sparseConvNetTensor import SparseConvNetTensor


class ValidConvolution(SparseModule):
    def __init__(self, dimension, nIn, nOut, filter_size, bias):
        SparseModule.__init__(self)
        self.dimension = dimension
        self.nIn = nIn
        self.nOut = nOut
        self.filter_size = toLongTensor(dimension, filter_size)
        self.filter_volume = self.filter_size.prod()
        self.weight = torch.Tensor(
            nIn * self.filter_volume, nOut
        ).normal_(0, (2.0 / nIn / self.filter_volume)**0.5)
        self.gradWeight = torch.Tensor(nIn * self.filter_volume, nOut)
        if bias:
            self.bias = torch.Tensor(nOut).zero_()
            self.gradBias = torch.Tensor(nOut)
        self.output = SparseConvNetTensor(torch.Tensor())
        self.gradInput = torch.Tensor()

    def updateOutput(self, input):
        assert input.features.size(1) == self.nIn
        self.output.metadata = input.metadata
        self.output.spatial_size = input.spatial_size
        s.forward_pass_multiplyAdd_count +=\
            dim_typed_fn(self.dimension, input, 'ValidConvolution_updateOutput')(
                input.spatial_size,
                self.filter_size,
                input.metadata.ffi,
                input.features,
                self.output.features,
                self.weight,
                optionalTensor(self, 'bias'),
                self.filter_volume,
                torch.cuda.IntTensor() if input.features.is_cuda else nullptr)
        s.forward_pass_hidden_states += self.output.features.nelement()
        return self.output

    def backward(self, input, gradOutput, scale=1):
        assert scale == 1
        dim_typed_fn(self.dimension, input, 'ValidConvolution_backward')(
            input.spatial_size,
            self.filter_size,
            input.metadata.ffi,
            input.features,
            self.gradInput,
            gradOutput,
            self.weight,
            self.gradWeight,
            optionalTensor(self, 'gradBias'),
            self.filter_volume,
            torch.cuda.IntTensor() if input.features.is_cuda else nullptr)
        return self.gradInput

    def type(self, t=None, tensorCache=None):
        if t is None:
            return self._type
        self._type = t
        self.weight = self.weight.type(t)
        self.gradWeight = self.gradWeight.type(t)
        self.output.type(t)
        self.gradInput = self.gradInput.type(t)
        if hasattr(self, 'bias'):
            self.bias = self.bias.type(t)
            self.gradBias = self.gradBias.type(t)

    def __repr__(self):
        s = 'ValidConvolution ' + str(self.nIn) + '->' + str(self.nOut) + ' C'
        if self.filter_size.max() == self.filter_size.min():
            s = s + str(self.filter_size[0])
        else:
            s = s + '(' + str(self.filter_size[0])
            for i in self.filter_size[1:]:
                s = s + ',' + str(i)
            s = s + ')'
        return s
